crates/iceberg/src/spec/values.rs

```diff
@@ -70,6 +79,571 @@ pub enum PrimitiveLiteral {
     Decimal(i128),
 }
 
+/// Literal associated with its type. The value and type pair is checked when construction, so the type and value is
+/// guaranteed to be correct when used.
+///
+/// By default we decouple the type and value of a literal, so we can use avoid the cost of storing extra type info
+/// for each literal. But associate type with literal can be useful in some cases, for example, in unbound expression.
+#[derive(Debug)]
+pub struct Datum {
+    r#type: PrimitiveType,
+    literal: PrimitiveLiteral,
+}
+
+impl Display for Datum {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match (&self.r#type, &self.literal) {
+            (_, PrimitiveLiteral::Boolean(val)) => write!(f, "{}", val),
+            (_, PrimitiveLiteral::Int(val)) => write!(f, "{}", val),
+            (_, PrimitiveLiteral::Long(val)) => write!(f, "{}", val),
+            (_, PrimitiveLiteral::Float(val)) => write!(f, "{}", val),
+            (_, PrimitiveLiteral::Double(val)) => write!(f, "{}", val),
+            (_, PrimitiveLiteral::Date(val)) => write!(f, "{}", days_to_date(*val)),
+            (_, PrimitiveLiteral::Time(val)) => write!(f, "{}", microseconds_to_time(*val)),
+            (_, PrimitiveLiteral::Timestamp(val)) => {
+                write!(f, "{}", microseconds_to_datetime(*val))
+            }
+            (_, PrimitiveLiteral::TimestampTZ(val)) => {
+                write!(f, "{}", microseconds_to_datetimetz(*val))
+            }
+            (_, PrimitiveLiteral::String(val)) => write!(f, "{}", val),
+            (_, PrimitiveLiteral::UUID(val)) => write!(f, "{}", val),
+            (_, PrimitiveLiteral::Fixed(val)) => display_bytes(val, f),
+            (_, PrimitiveLiteral::Binary(val)) => display_bytes(val, f),
+            (
+                PrimitiveType::Decimal {
+                    precision: _,
+                    scale,
+                },
+                PrimitiveLiteral::Decimal(val),
+            ) => {
+                write!(f, "{}", Decimal::from_i128_with_scale(*val, *scale))
+            }
+            (_, _) => {
+                unreachable!()
+            }
+        }
+    }
+}
+
+fn display_bytes(bytes: &[u8], f: &mut Formatter<'_>) -> std::fmt::Result {
+    let mut s = String::with_capacity(bytes.len() * 2);
+    for b in bytes {
+        s.push_str(&format!("{:02X}", b));
+    }
+    f.write_str(&s)
+}
+
+impl From<Datum> for Literal {
+    fn from(value: Datum) -> Self {
+        Literal::Primitive(value.literal)
+    }
+}
+
+impl Datum {
+    /// Creates a boolean value.
```

Comment:
Great work on all the comment sections, they really help to place the pieces into context 🙌